### ISSUE TITLE
Failures of datetime conversions

### ISSUE THREAD

**Issue by impredicative** [ritchie46/polars#476]
1. There isn't an implicit way to convert Pandas datetime objects to Polars. I currently have to first convert relevant Pandas datetimes columns to Pandas strings, the Pandas strings to Polars strings, and finally the Polars strings to Polars Date64.
2. A Polars Date64 column having **datetime** data (not just date time) loses its **time** component altogether when converted to Numpy or to Arrow.

In effect I see no good way of datetimes into and out of Polars. This is a major blocker and makes Polars completely unusable for me.

**Comment by impredicative** [comment]
Basically I see no way of using the various Polars Time datatypes even though these datatypes do exist. I tried `series.cast` and it failed every time.

**Comment by ritchie46** [comment]
Can you give me a bit more context of what the exact datetime objects are in pandas? Currently `datetime64[ns]` is converted to `Date64` when converting from pandas df to polars.

*snippet used in conversion*

```python
        elif dtype == "datetime64[ns]":
            data[name] = pa.compute.cast(
                pa.array(np.array(df[name].values, dtype="datetime64[ms]")), pa.date64()
            )
```

> A Polars Date64 column having datetime data (not just date time) loses its time component altogether when converted to Numpy or to Arrow.

This I don't really understand. Polars `Date64`is an arrow `Date64`? Is the the polars data type accidentally converted to a different datatype?

**Comment by ritchie46** [comment]
Hmm.. Maybe the cast from ns to ms is wrong. :thinking:

**Comment by impredicative** [comment]
> Currently datetime64[ns] is converted to Date64 when converting from pandas df to polars.

The following code sample raises `ArrowInvalid: Timestamp value had non-zero intraday milliseconds` in its last line:

```
import datetime
import pandas as pd
import polars as pl

pl.from_pandas(pd.Series(['aa', 'bb']).to_frame('col1'))  # This example works.

pl.from_pandas((pd.Series([datetime.datetime.today(), datetime.datetime.now()]).to_frame('dt1')))  # This raises ArrowInvalid.
```

Also, `Date64` is extremely confusing if it can store datetime values but it doesn't export times at all.

**Comment by ritchie46** [comment]
Thanks for the snippet. I've been taking a look. It seems that casting to pyarrow's `Date64` always drops the time component, which is clearly not wanted. I see if we can utilize polars native casting for the `Date64` dtype.

**Comment by impredicative** [comment]
Before an issue is closed, I reasonably request:

* Sample code showing the change in behavior
* An actual release to PyPI

Also, it is not clear if Date64 will export times to numpy, parquet, etc.

**Comment by ritchie46** [comment]
>     * Sample code showing the change in behavior

Check the added test in a MR. 

> Also, it is not clear if Date64 will export times to numpy, parquet, etc.

The time component was lost when casting to date64. This should now be fixed.


> An actual release to PyPI

Sorry, but I will not do that. The tip of the master shows the state related to issues. I have got also a Rust library which release cycle is not synced with the Python bindings.

**Comment by impredicative** [comment]
> Check the added test in a MR.

I see one test. I see no test for a successful datetime export to numpy and arrow.

**Comment by impredicative** [comment]
@ritchie46 Point 2 of this issue is completely ignored. I have been trying to tell you over and over that the conversion of the time component to numpy and arrow isn't working. These details just get ignored, and the issue prematurely closed. I guess I'll just move on then.

**Comment by ritchie46** [comment]
[You have a habit of complaining](https://github.com/ritchie46/polars/issues/401) and being very unfriendly on a FOSS repo. You tell me things are "cleary very obvious" without delivering any valuable contributions. TBH, I am glad to see you move on.

**Comment by impredicative** [comment]
I strongly get the impression that you get paid for closing issues, with or with without a resolution. I never ask for any fancy features, only for what is expected dataframe functionality. In the case of the current issue, bugs were reported and repeatedly ignored. This project is unsuitable for any serious work, and will be so for a long time.

**Comment by bwlewis** [comment]
Sorry to ask about a closed issue, but I'm also having trouble converting Pandas data frames with datetime[ns] values to polars. I tried the following minimal example with the current GitHub repo build (Version: 0.13.12) and I can't figure this out.

```
import pandas as pd
import polars as p

d = pd.DataFrame([pd.Timestamp(year = 2021,  month = 1, day = 1, hour = 1, second = 1, nanosecond = 1)], columns=["date"])
print(d.dtypes)
print(p.convert.from_pandas(d))
```

As you can see the resulting polars data appears as datetime[ms]. Is there some easy work-around? Any help would be appreciated.

thanks!

**Comment by bwlewis** [comment]
Here is a (crude) resolution to my previous comment, sufficient for me for now (I am certain this can be improved though):

```
import pandas as pd
import polars as p
d = pd.DataFrame([pd.Timestamp(year = 2021,  month = 1, day = 1, hour = 1, second = 1, nanosecond = 1)], columns=["date"])
print(d)
print(p.convert.from_pandas(d))
print((p.convert.from_pandas(d)).to_pandas() == d)

# resolution
d["date"] = d["date"].astype(int)
d_plr = p.convert.from_pandas(d)
d_plr["date"] = d_plr["date"].cast(p.Datetime).dt.and_time_unit("ns")
print(d_plr)
```

**Comment by bwlewis** [comment]
FYI, I was using convert.from_pandas for a broad comparison of various 'as-of' joins here:  https://bwlewis.github.io/duckdb_and_r/asof/asof.html

Polars performance is particularly mind-blowingly-good in this example! :1st_place_medal:

**Comment by ritchie46** [comment]
Thanks for pointing this out. This is actually a bug on our side. I fixed it in #2936. This will be released this afternoon.

Note that it is intended to access `from_pandas` directly from the root of the polars namespace like so:

```python
import polars as pl
from datetime import datetime

df = pd.DataFrame([pd.Timestamp(year = 2021,  month = 1, day = 1, hour = 1, second = 1, nanosecond = 1)], columns=["date"])
assert pl.from_pandas(df)[0, 0] == datetime(2021, 1, 1, 1, 0, 1)
```

> Polars performance is particularly mind-blowingly-good in this example! 1st_place_medal

Awesome! Have never benchmarked the asof join, so it is good to know we are rocking it. :smile:
